pkg/tsdb/cloudwatch/metric_find_query.go

```diff
@@ -631,36 +638,31 @@ func (e *cloudWatchExecutor) handleGetResourceArns(ctx context.Context, paramete
 	return result, nil
 }
 
-func (e *cloudWatchExecutor) cloudwatchListMetrics(region string, namespace string, metricName string,
-	dimensions []*cloudwatch.DimensionFilter) (*cloudwatch.ListMetricsOutput, error) {
-	svc, err := e.getCWClient(region)
+func (e *cloudWatchExecutor) listMetrics(region string, params *cloudwatch.ListMetricsInput) ([]*cloudwatch.Metric, error) {
+	client, err := e.getCWClient(region)
 	if err != nil {
 		return nil, err
 	}
 
-	params := &cloudwatch.ListMetricsInput{
-		Namespace:  aws.String(namespace),
-		Dimensions: dimensions,
-	}
+	plog.Debug("Listing metrics pages")
+	cloudWatchMetrics := []*cloudwatch.Metric{}
 
-	if metricName != "" {
-		params.MetricName = aws.String(metricName)
-	}
+	pageNum := 0
+	err = client.ListMetricsPages(params, func(page *cloudwatch.ListMetricsOutput, lastPage bool) bool {
+		pageNum++
+		metrics.MAwsCloudWatchListMetrics.Inc()
+		metrics, err := awsutil.ValuesAtPath(page, "Metrics")
+		if err != nil {
+			return !lastPage && pageNum < e.settings.ListMetricsPageLimit
+		}
 
-	var resp cloudwatch.ListMetricsOutput
-	if err := svc.ListMetricsPages(params,
-		func(page *cloudwatch.ListMetricsOutput, lastPage bool) bool {
-			metrics.MAwsCloudWatchListMetrics.Inc()
-			metrics, _ := awsutil.ValuesAtPath(page, "Metrics")
-			for _, metric := range metrics {
-				resp.Metrics = append(resp.Metrics, metric.(*cloudwatch.Metric))
-			}
-			return !lastPage
-		}); err != nil {
-		return nil, fmt.Errorf("failed to call cloudwatch:ListMetrics: %w", err)
-	}
+		for _, metric := range metrics {
+			cloudWatchMetrics = append(cloudWatchMetrics, metric.(*cloudwatch.Metric))
+		}
+		return !lastPage && pageNum < e.settings.ListMetricsPageLimit
```

Comment:
Maybe something like:

```
		if err == nil {
			for _, metric := range metrics {
				cloudWatchMetrics = append(cloudWatchMetrics, metric.(*cloudwatch.Metric))
			}
		}
		return !lastPage && pageNum < e.settings.ListMetricsPageLimit
```

to avoid duplicating return statement?

Reply:
yes much better!